app/bundles/LeadBundle/Tests/Entity/TagRepositoryTest.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * @copyright   2016 Mautic Contributors. All rights reserved
+ * @author      Mautic, Inc.
+ *
+ * @link        https://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Tests\Entity;
+
+use Doctrine\ORM\EntityManager;
+use Mautic\LeadBundle\Entity\Tag;
+use Mautic\LeadBundle\Entity\TagRepository;
+
+class TagRepositoryTest extends \PHPUnit_Framework_TestCase
+{
+    public function testSaveEntityWithUniqueNewTag()
+    {
+        $entity = new Tag('sometag');
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->once())
+            ->method('persist')
+            ->with($entity);
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $mockRepository->expects($this->once())
+            ->method('findOneBy')
+            ->with(['tag' => 'sometag'])
+            ->willReturn(null);
+
+        $mockRepository->saveEntity($entity);
+
+        $this->assertNull($entity->getId());
+    }
+
+    public function testSaveEntityWithExistingTag()
+    {
+        $entity = new Tag('sometag');
+        $entity->setId(45);
+
+        $mockEntityManager = $this->getMockBuilder(EntityManager::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['persist', 'flush'])
+            ->getMock();
+
+        $mockEntityManager->expects($this->once())
+            ->method('persist')
+            ->with($entity);
+
+        $mockRepository = $this->getMockBuilder(TagRepository::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getEntityManager', 'findOneBy'])
+            ->getMock();
+
+        $mockRepository->method('getEntityManager')
+            ->willReturn($mockEntityManager);
+
+        $mockRepository->expects($this->never())
+            ->method('findOneBy');
+
+        $mockRepository->saveEntity($entity);
+
+        $this->assertSame(45, $entity->getId());
```

Comment:
This test does not make much sense - You have set ID manually, so why to assert that is same? There is no code to change it.

Reply:
As you noted above, the saveEntity method replaces the entity if it doesn't have ID but the tag exists. This test ensures it works correctly for entities which does have ID. I'm just covering all the expected possibilities with those tests. I don't see anything wrong with it.